src/tiled/tileseteditor.cpp

```diff
@@ -465,7 +465,28 @@ void TilesetEditor::addTiles()
         Tile *newTile = new Tile(tileset->takeNextTileId(), tileset);
         newTile->setImage(loadedFile.image);
         newTile->setImageSource(loadedFile.imageSource);
-        tiles.append(newTile);
+
+        // If the tile is already in the tileset, warn user and confirm addition
+        bool addImage = true;
+        const QMap<int, Tile*> addedTiles = mCurrentTilesetDocument->tileset().data()->tiles();
+        for(auto tile : addedTiles) {
```

Comment:
At this point we have already loaded all images the user is trying to add, which is the slowest part of the operation. Given that one use-case for this feature is that you can pick a large collection of images and have Tiled figure out which ones still need to be added, it would be good to do this duplicate check in the previous loop that iterated the chosen files.